File: app/src/main/java/com/ps/simplepapersoccer/Activities/GameView.java
package com.ps.simplepapersoccer.Activities;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.ScaleDrawable;
import android.util.AttributeSet;
import android.view.View;

import com.ps.simplepapersoccer.GameObjects.LinesToDraw;
import com.ps.simplepapersoccer.GameObjects.Move.PossibleMove;
import com.ps.simplepapersoccer.GameObjects.Node;
import com.ps.simplepapersoccer.GameObjects.Player;
import com.ps.simplepapersoccer.R;

public class GameView extends View {

 	public List<LinesToDraw> drawLines = new ArrayList<LinesToDraw>();
	private GameActivity gameActivity;
    
	private Canvas canvas;
	private Paint paint;

	private int nodeSize = 20;
	
	public float leftEdge = 50;
	public float topEdge = 100;
	
	public float gridXdraw;
	public float gridYdraw;
	
	public int gridSizeX = 8;
	public int gridSizeY = 10;

	private int gridStrokeWidth = 4;
	private int gridColor = Color.rgb(242, 244, 247);

	public float leftEdgeMargin = 30;
	public float rightEdgeMargin = 1.027f;

	private int ballSize = 100;
	private int sideLineStrokeWidth = 8;
	
	public float rightEdge;
	public float bottomEdge;
	
	protected float middlePointX;
	protected float middlePointY;
	
	private Player playerTurn;

	private float ballX;
	private float ballY;
		
	public GameView(Context context)
	{
		super(context);
	}
	
    public GameView(Context context, AttributeSet attrs)
	{
        super(context, attrs);
    }

	public GameView(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
	}
    
    public void UpdateBallPosition(float[] ballCoords, Player playerTurn)
    {
    	this.ballX = ballCoords[0];
    	this.ballY = ballCoords[1];
    	this.playerTurn = playerTurn;
    }
    
    public void SetValues(int screenWidth, int screenHeight, int gridSizeX, int gridSizeY, GameActivity gameActivity)
    {
        topEdge = screenHeight/9;
		leftEdge = screenWidth/leftEdgeMargin;
		
		this.gridSizeX = gridSizeX;
		this.gridSizeY = gridSizeY;
		
		bottomEdge = screenHeight/1.2f;
		rightEdge = screenWidth/rightEdgeMargin;
			
		gridXdraw = (rightEdge-leftEdge)/gridSizeX;
		gridYdraw = (bottomEdge-topEdge)/gridSizeY;
		
		this.middlePointX = (rightEdge+leftEdge)/2;
		this.middlePointY = (bottomEdge+topEdge)/2;

		this.gameActivity = gameActivity;
		
		this.invalidate();
    }

	@Override
    protected void onDraw(Canvas canvas) {
	        super.onDraw(canvas);
	        this.canvas = canvas;		   
	        RedrawMap();
    }
    
    public void RedrawMap()
    {
        paint = new Paint();
        paint.setTextSize(40);
		paint.setStrokeWidth(gridStrokeWidth);
    	paintComponent(canvas, paint);
    }
    
    public void paintComponent(Canvas canvas, Paint paint) {
		//Clear before repaint
		canvas.clipRect(0, 0, getWidth(), getHeight());
		
		this.middlePointX = (rightEdge+leftEdge)/2;
		this.middlePointY = (bottomEdge+topEdge)/2;

		paint.setColor(gridColor);
		for(float outerLoop = 0; outerLoop <= gridSizeX; ++outerLoop)
		{
			for (float innerLoop = 1; innerLoop < gridSizeY; ++innerLoop)
			{
				//Vert
				canvas.drawLine(outerLoop*gridXdraw+leftEdge, topEdge+gridYdraw, outerLoop*gridXdraw+leftEdge, bottomEdge-gridYdraw, paint);

				//Horz
				canvas.drawLine(leftEdge, innerLoop*gridYdraw+topEdge, rightEdge, innerLoop*gridYdraw+topEdge, paint);
			}
		}

		DrawGoal(topEdge, paint, Color.RED);
		DrawGoal(bottomEdge, paint, Color.BLUE);

		DrawPossibleMoves(paint);
		
		paint.setColor(Color.BLACK);
		paint.setStrokeWidth(sideLineStrokeWidth);

		DrawSideline(leftEdge, paint);
		DrawSideline(rightEdge, paint);

		DrawGoalLines(topEdge, paint, topEdge+gridYdraw, Color.RED);
		DrawGoalLines(bottomEdge, paint, bottomEdge-gridYdraw, Color.BLUE);
		
		paint.setStyle(Paint.Style.FILL);
		
		for(int i = 0; i < drawLines.size(); ++i)
		{
			paint.setColor(drawLines.get(i).color);
			canvas.drawLine(drawLines.get(i).fromX,drawLines.get(i).fromY, drawLines.get(i).toX,drawLines.get(i).toY, paint);
		}

		//Draw ball
		Drawable image = getResources().getDrawable(R.drawable.football);
		image = new ScaleDrawable(image, 0, ballSize, ballSize).getDrawable();
		image.setBounds((int)ballX-ballSize/2, (int)ballY-ballSize/2, (int)ballX+ballSize/2, (int)ballY+ballSize/2);
		image.draw(canvas);
	}

	private void DrawPossibleMoves(Paint paint)
	{
		paint.setColor(gameActivity.gameHandler.currentPlayersTurn.playerColor);
		if (gameActivity.gameHandler.aiTurn) return;
		for (PossibleMove move : gameActivity.gameHandler.allPossibleMovesFromNode(gameActivity.gameHandler.ballNode)) {
			float[] coords = gameActivity.nodeToCoords(move.newNode);
			canvas.drawCircle(coords[0], coords[1], 20, paint);
		}
	}

	private void DrawGoalLines(float edge, Paint paint, float edgeMargin, int color)
	{
		canvas.drawLine(leftEdge, edgeMargin, middlePointX-gridXdraw, edgeMargin, paint);
		canvas.drawLine(rightEdge, edgeMargin, middlePointX+gridXdraw, edgeMargin, paint);

		canvas.drawLine(middlePointX+gridXdraw, edge, middlePointX+gridXdraw, edgeMargin, paint);
		canvas.drawLine(middlePointX-gridXdraw, edge, middlePointX+gridXdraw, edge, paint);
		canvas.drawLine(middlePointX-gridXdraw, edgeMargin, middlePointX-gridXdraw, edge, paint);

		paint.setColor(color);
		paint.setStyle(Paint.Style.STROKE);
		paint.setColor(Color.BLACK);
	}

	private void DrawGoal(float edge, Paint paint, int color)
	{
		paint.setColor(color);
		canvas.drawCircle(middlePointX, edge, nodeSize, paint);
	}

	private void DrawSideline(float edge, Paint paint)
	{
		canvas.drawLine(edge, topEdge+gridYdraw, edge, bottomEdge-gridYdraw, paint);
	}
}

File: app/src/main/java/com/ps/simplepapersoccer/Activities/GameView.kt
package com.ps.simplepapersoccer.activities

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import com.ps.simplepapersoccer.gameObjects.game.geometry.TwoDimensionalPoint
import android.graphics.drawable.ScaleDrawable
import android.util.AttributeSet
import android.view.View
import com.ps.simplepapersoccer.data.enums.NodeTypeEnum
import com.ps.simplepapersoccer.gameObjects.game.GameBoard
import com.ps.simplepapersoccer.gameObjects.game.GameViewDrawData
import com.ps.simplepapersoccer.gameObjects.game.geometry.abstraction.IntegerLine
import com.ps.simplepapersoccer.gameObjects.game.geometry.LinesToDraw
import com.ps.simplepapersoccer.gameObjects.game.geometry.Node
import com.ps.simplepapersoccer.R
import com.ps.simplepapersoccer.viewmodel.GameViewModel
import kotlin.collections.HashSet

class GameView : View {

    var drawLines: HashSet<LinesToDraw> = HashSet()

    var gameViewDrawData: GameViewDrawData? = null

    private var gameViewModel: GameViewModel? = null
    private var gameBoard: GameBoard? = null

    private var canvas: Canvas? = null
    private var paint = Paint()

    private var leftEdge = 50f
    private var topEdge = 100f

    private var gridXdraw: Float = 0.toFloat()
    private var gridYdraw: Float = 0.toFloat()

    var gridSizeX = 0
    var gridSizeY = 0

    private val gridStrokeWidth = 4
    private val gridColor = Color.rgb(242, 244, 247)

    private var leftEdgeMargin = 30f
    private var rightEdgeMargin = 1.027f

    private val ballSize = 100
    private val sideLineStrokeWidth = 8

    private var rightEdge: Float = 0.toFloat()
    private var bottomEdge: Float = 0.toFloat()

    private var middlePointX: Float = 0.toFloat()
    private var middlePointY: Float = 0.toFloat()

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

    fun init(screenWidth: Int, screenHeight: Int, gridSizeX: Int, gridSizeY: Int, gameViewModel: GameViewModel, gameBoard: GameBoard) {
        topEdge = (screenHeight / 9).toFloat()
        leftEdge = screenWidth / leftEdgeMargin

        this.gridSizeX = gridSizeX
        this.gridSizeY = gridSizeY

        bottomEdge = screenHeight / 1.2f
        rightEdge = screenWidth / rightEdgeMargin

        gridXdraw = (rightEdge - leftEdge) / gridSizeX
        gridYdraw = (bottomEdge - topEdge) / gridSizeY

        this.middlePointX = (rightEdge + leftEdge) / 2
        this.middlePointY = (bottomEdge + topEdge) / 2

        this.gameViewModel = gameViewModel
        this.gameBoard = gameBoard

        this.invalidate()
    }

    //Converts node coordinates to screen coordinates
    fun nodeToCoords(n: Node): FloatArray {
        val coords = FloatArray(2)
        coords[0] = n.getVisibleCoords.x * gridXdraw + leftEdge
        coords[1] = n.getVisibleCoords.y * gridYdraw + topEdge
        return coords
    }

    private fun pointsCoordsToCoords(point: TwoDimensionalPoint): FloatArray {
        val newCoords = FloatArray(2)
        newCoords[0] = point.x / 2 * gridXdraw + leftEdge
        newCoords[1] = point.y / 2 * gridYdraw + topEdge
        return newCoords
    }

    //Converts screen coordinates to node coordinates
    fun coordsToNode(argX: Float, argY: Float): FloatArray {
        val x = (argX + gridXdraw.div(2)) / gridXdraw * gridXdraw
        val y = (argY + gridYdraw.div(2)) / gridYdraw * gridYdraw - topEdge

        val coords = FloatArray(2)
        coords[0] = x / gridXdraw
        coords[1] = y / gridYdraw
        return coords
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        this.canvas = canvas
        redrawMap()
    }

    fun drawAsync(gameViewDrawData: GameViewDrawData){
        drawLines.add(gameViewDrawData.drawLine!!)
        this.gameViewDrawData = gameViewDrawData
        gameViewModel?.updatePlayerTurnText()
        this.invalidate()
    }

    private fun redrawMap() {
        paint = Paint()
        paint.textSize = 40f
        paint.strokeWidth = gridStrokeWidth.toFloat()
        paintComponent(canvas!!, paint)
    }

    private fun paintComponent(canvas: Canvas, paint: Paint) {
        //Clear before repaint
        canvas.clipRect(0, 0, width, height)

        this.middlePointX = (rightEdge + leftEdge) / 2
        this.middlePointY = (bottomEdge + topEdge) / 2

        paint.color = gridColor

        for (node in gameBoard?.nodesHashSet!!) {
            node.coordNeighbors
                    .filterNot { node.coords.x != it.coords.x && node.coords.y != it.coords.y }
                    .forEach { canvas.drawLine(nodeToCoords(node)[0], nodeToCoords(node)[1], nodeToCoords(it)[0], nodeToCoords(it)[1], paint) }
        }

        drawGoalLine(topEdge, paint, Color.RED, gameBoard!!.topGoalLines.goalLine)
        drawGoalLine(bottomEdge, paint, Color.BLUE, gameBoard!!.bottomGoalLines.goalLine)

        paint.color = Color.BLACK
        paint.strokeWidth = sideLineStrokeWidth.toFloat()

        drawWalls(paint)
        drawGoalPosts(paint, gameBoard!!.topGoalLines.leftPost, gameBoard!!.topGoalLines.rightPost)
        drawGoalPosts(paint, gameBoard!!.bottomGoalLines.leftPost, gameBoard!!.bottomGoalLines.rightPost)

        paint.style = Paint.Style.FILL

        drawLines.forEach {
            paint.color = it.color
            canvas.drawLine(it.fromPoint.x, it.fromPoint.y, it.toPoint.x, it.toPoint.y, paint)
        }

        drawPossibleMoves()

        var image = resources.getDrawable(R.drawable.football)
        image = ScaleDrawable(image, 0, ballSize.toFloat(), ballSize.toFloat()).drawable
        val ballNodeCoords = pointsCoordsToCoords(gameViewDrawData?.ballNode?.coords!!)
        image.setBounds(ballNodeCoords[0].toInt() - ballSize / 2, ballNodeCoords[1].toInt() - ballSize / 2, ballNodeCoords[0].toInt() + ballSize / 2, ballNodeCoords[1].toInt() + ballSize / 2)
        image.draw(canvas)
    }

    private fun drawPossibleMoves() {
        paint.color = gameViewDrawData?.currentPlayerTurn?.playerColor as Int
        if (gameViewDrawData?.currentPlayerTurn?.isAi!!) return
        gameViewDrawData?.nodeNeighbors!!
                .map { nodeToCoords(it) }
                .forEach {
                    canvas?.drawCircle(it[0], it[1], 20f, paint)
                }
    }

    private fun drawGoalLine(edge: Float, paint: Paint, color: Int, line: IntegerLine) {
        paint.color = color
        paint.strokeWidth = sideLineStrokeWidth.toFloat()
        val startPoint = pointsCoordsToCoords(line.fromPoint)
        val endPoint = pointsCoordsToCoords(line.toPoint)
        canvas?.drawLine(startPoint[0], startPoint[1], endPoint[0], endPoint[1], paint)
    }

    private fun drawGoalPosts(paint: Paint, leftLine: IntegerLine, rightLine: IntegerLine){
        var startPoint = pointsCoordsToCoords(leftLine.fromPoint)
        var endPoint = pointsCoordsToCoords(leftLine.toPoint)
        canvas?.drawLine(startPoint[0], startPoint[1], endPoint[0], endPoint[1], paint)
        startPoint = pointsCoordsToCoords(rightLine.fromPoint)
        endPoint = pointsCoordsToCoords(rightLine.toPoint)
        canvas?.drawLine(startPoint[0], startPoint[1], endPoint[0], endPoint[1], paint)
    }

    private fun drawWalls(paint: Paint) {
        gameBoard?.nodesHashSet?.filter { otherNode -> otherNode.nodeType == NodeTypeEnum.Wall }?.forEach{
            it.coordNeighbors.filter { otherNode -> otherNode.nodeType == NodeTypeEnum.Wall }.forEach {
                otherNode -> if (!it.isDiagonalNeighbor(otherNode)) canvas?.drawLine(nodeToCoords(it)[0], nodeToCoords(it)[1], nodeToCoords(otherNode)[0], nodeToCoords(otherNode)[1], paint)
            }
        }

        gameBoard?.nodesHashSet?.filter { otherNode -> otherNode.nodeType == NodeTypeEnum.Wall }?.forEach{
            it.coordNeighbors.filter { otherNode -> otherNode.nodeType == NodeTypeEnum.Post }.forEach{
                otherNode -> if (!it.isDiagonalNeighbor(otherNode)) canvas?.drawLine(nodeToCoords(it)[0], nodeToCoords(it)[1], nodeToCoords(otherNode)[0], nodeToCoords(otherNode)[1], paint)
            }
        }
    }
}
